### ISSUE TITLE
[DO - enrichment] variable references in aggregation

### ISSUE THREAD

**Issue by oleurud** [CartoDB/cartoframes#1174]
Comes from https://github.com/CartoDB/cartoframes/pull/1140#issuecomment-549366367

One thing I find confusing is how the agg_operators part works. When there are no aggregation operators specified it's pretty clear (use the default, if available in the metadata agg_method field).

However, I find it weird that the user has to refer to a variable by slug (or ID, I suppose) but then specify the aggregation using the variable's column_name as the dictionary key. I think this has several drawbacks:

- I understand we specify variables to enrich (instead of full datasets) to allow enriching with variables from different datasets. But what if the variables have the same column_name in their respective datasets? For example, if I were to enrich my geometry with two population columns from different datasets (both called `pop), in order to compare them...

- If I already know the slug (which uniquely identifies a variable), why not just use that for the agg_operators dict, too? Of course, that would imply looking up the proper column ID in the metadata (but we're already doing that).

- It'd be nice to be able to specify several aggregation functions for a single variable (just like Pandas does). In this case, we'd have to differentiate them somehow (maybe adding a {func}_ prefix?)

**Comment by oleurud** [comment]
Thinking about it, my first idea was:
- use Variable instances or Variables id/slug (in the same way we are doing in variables parameter) instead of the column name 

But a new approach comes to my mind: what about using `variables` to define the aggregations too? The variables would receive a list of:
- **variables or variables ids**: In this case, the aggregation method would be the default one defined in the catalog Variable (or the default one (`array_agg`) if the variable has not aggregation method defined)
```
variables = [Variable1 instance, Variable2 instance, ...]
variables = [variable1 id, variable2 id, ...]
```
- **a variable or variable id and the aggregation method pairs**. With that, the user can overwrite the default aggregation method:
```
v1 = {
    'variable': Variable1,
    'aggregation': 'SUM' 
}
variables = [v1, ...]
```

What do you think @alasarr @arredond @alejandrohall @cmongut ?

**Comment by elenatorro** [comment]
From an API perspective, I see a couple things (not sure if they makes sense, just writing my thoughts here):

First proposal:

- Declaring variables as `Dicts` is one of the approaches we started with (Layers, Popups...) but we're currently using the convention of creating classes.

- It might be confusing to have two types of variables: The ones in the Dataset and the ones we want to define for the aggregation. 

Having this in mind, I like the idea of having a class for this, but I think it should be something like:

```py
variables = [
    AggVariable(variable1.slug, 'SUM'),
    AggVariable(variable2.slug, 'AVG')
]
```

Second proposal:

I've a question: In this case, can the agg method of a variable be overwritten?

**Comment by oleurud** [comment]
Thank you, Elena :heart: 

First of all, the idea was to support both "proposals" not only one. So,
> Second proposal:: I've a question: In this case, can the agg method of a variable be overwritten?

In this case, the aggregation method would be the default one defined in the catalog Variable. So, if the catalog variable has the aggregation method, it will be used. If not, the default one (`array_agg`) will be used 

> It might be confusing to have two types of variables: The ones in the Dataset and the ones we want to define for the aggregation.

This is what I want to solve. the idea is to remove the `agg_operators`. I am going to write it in the previous comment.

About the classes usage, I am not a fan, but I understand the point and I see the advantage. Also, it makes sense doing in the same way in different parts of the code. So +1

**Comment by elenatorro** [comment]
Great! We don't need to use the classes explicitly, it could be (I'm using the slug, but it can also be the 'id' or the `variable` itself as you mentioned previously):

```py
variables = [[variable1.slug, 'SUM'], [variable2.slug, 'AVG']]
```

And, if taking default agg values:

* In one variable: (for example, in `variable2`):

```py
variables = [[variable1.slug, 'SUM'], variable2.slug]
```

* In both variables:

```py
variables = [variable1.slug, variable2.slug]
```

**Comment by oleurud** [comment]
^ yeah! that's is perfect IMO

**Comment by oleurud** [comment]
Moving forward!

**Comment by arredond** [comment]
I like this solution! Ordered but clean, IMHO

**Comment by oleurud** [comment]
I realized there is a case difficult to support using only one parameter (I did not realize that that cases exists).

Talking about polygons enrichment, the idea is to intersect polygons from client data with the polygons from DO. Thinking about the following (and usual) case: a polygon from the client intersects with 3 polygons from DO. Here, there are 2 cases: 
- do the aggregation in Bigquery: we return a row for each source polygon with the data aggregated
- do it manually locally: we return as many rows as polygons intersect (3 in the example) for each source polygon.

This second case is what I did not take into account, basically because that case is not working in the current version and I did not see the goal (why does a client want to do it herself? to use other aggregations not supported by BQ, to use custom aggregation methods, use other user data to do the aggregation, ...)

I don't want to spend more time on this issue, so I am going to solve the initial problem of the issue, using 2 different parameters and trying to do the aggregation one more verbose (for a better understanding)
